### PR TITLE
WIP: Load scrape configs from multiple files

### LINKED ISSUE #8543: Load scrape configs from multiple files
This issue is about providing the ability to split Prometheus scrape configs into multiple files.

Design doc: https://docs.google.com/document/d/1bSehbL_Rzx9_guVgISmara-Ft_FB57q_YxEZtwxE3C8/edit#heading=h.bupciudrwmna

Add a new parameter `scrape_config_files: [ - <filepath_glob> ... ]`.

The files in the scrape_config_files are read, and may contain one or multiple files. The filenames can contain globs (scrape_configs.d/*.yml). Relative paths are relative to the main config file.

The files are re-read on every reload. Wrong scrape configs make the reload fail, and scrape configs as applied before apply. Errors reading scrape configs prevent Prometheus to start.

Scrape config names must be unique between prometheus.yml and imported files, otherwise reload/start fail.

Scrape configs are never merged together.

Exemple scrape config file:

Prometheus.yml

```
global: 
   scrape_interval: 10s
scrape_configs:
  - job_name: 'prometheus'
    static_configs:
    - targets: ['localhost:9090']
scrape_config_files:
  - blackbox.yml
```

blackbox.yml

```
scrape_configs:
 - job_name: 'blackbox'
    metrics_path: /probe
    params:
      module: [http_2xx]
    static_configs:
      - targets:
        - http://prometheus.io
        - https://prometheus.io
        - http://example.com:8080
    relabel_configs:
      - source_labels: [__address__]
        target_label: __param_target
      - source_labels: [__param_target]
        target_label: instance
      - target_label: __address__
        replacement: 127.0.0.1:9115  # The blackbox exporter's real hostname:port.
```

### PR DESCRIPTION
**Pull request by srikanthccv**
Resolves #8543

### CODE REVIEW

**Comment by lonewolf3739** [comment]
@roidelapluie This is missing docs update right now. I would appreciate pre-review meanwhile about the overall approach and if there is anything else missing.

File: config/config.go [hunk 1]
```diff
@@ -158,22 +162,90 @@ var (
...
+	var scrapeConfigFiles []string
+	for _, pat := range c.ScrapeConfigFiles {
+		fs, err := filepath.Glob(pat)
+		if err != nil {
+			// The only error can be a bad pattern.
+			return errors.Wrapf(err, "error retrieving scrape config files for %s", pat)
+		}
+		// There is no scrape config file that matches pattern.
+		if len(fs) == 0 {
+			return errors.Errorf("no scrape config files found for %s", pat)
```
Comment (roidelapluie) [inline config/config.go:192]: We should not error here; as the path is a glob, there could be no files.

**Review by roidelapluie — COMMENTED** [review]
Thank you for this! I did not have the time to come to this today, I will check this in the next 24 hours. Thanks!

File: config/config.go (hunk 1 above)
Comment (lonewolf3739) [reply config/config.go:192]: My line of thought was `filepath.Glob` returns err only when pattern is bad, If they provide a valid pattern and there are no files then it probably means there was a misspell or typo. What would be the ideal behaviour here? Should we err out here or skip such cases?

File: config/config.go [hunk 2]
```diff
@@ -158,22 +162,90 @@ var (
...
+		for _, scfg := range scfgFile.ScrapeConfigs {
+			err := c.checkScrapeConfig(scfg)
+			if err != nil {
+				return err
+			}
+			if _, ok := jobNames[scfg.JobName]; ok {
+				return errors.Errorf("found multiple scrape configs with job name %q", scfg.JobName)
+			}
+			jobNames[scfg.JobName] = struct{}{}
+			c.ScrapeConfigs = append(c.ScrapeConfigs, scfg)
```
Comment (roidelapluie) [inline config/config.go:217]: I would not mutate the main configuration, but rather refactor the logic to pass a complete list of scrape configs to the scrape config manager.

File: config/config.go [hunk 3]
```diff
@@ -158,22 +162,90 @@ var (
...
+	content, err := ioutil.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	// Parse content into ScrapeConfigFile.
+	cfg := &ScrapeConfigFile{}
+	err = yaml.UnmarshalStrict([]byte(content), cfg)
+	if err != nil {
+		return nil, err
+	}
```
Comment (roidelapluie) [inline config/config.go:235]: You should call SetDirectory() here.

File: config/config.go (hunk 1 above)
Comment (roidelapluie) [reply config/config.go:192]: This should be silently ignored if there  are no files, to match what we do for rule files.

File: config/config.go [hunk 4]
```diff
@@ -158,22 +162,90 @@ var (
...
+		if err != nil {
+			return err
+		}
+		for _, scfg := range scfgFile.ScrapeConfigs {
+			err := c.checkScrapeConfig(scfg)
+			if err != nil {
+				return err
+			}
+			if _, ok := jobNames[scfg.JobName]; ok {
+				return errors.Errorf("found multiple scrape configs with job name %q", scfg.JobName)
```
Comment (roidelapluie) [inline config/config.go:214]: It seems we should do this check just once, and not one time for the main file and one time for the secondary files.

**Review by roidelapluie — COMMENTED** [review]
Thank you very much for your work.

As this is a draft, I have done a high-level review of the design, and what I think is the most annoying to me is that we should not mutate the original configuration because it would be confusing in the UI, but instead pass the an indiviual scrape job list to the scrape manager.

**Comment by DanielQujun** [comment]
@lonewolf3739 @roidelapluie  Hi, 
notice this pr has not been updated for a long time, are we still working on this? I'm really looking forward to this feature, If @lonewolf3739 is too busy for this, I'd like to continue this.

**Comment by lonewolf3739** [comment]
Apologies, I completely forgot about this. @DanielQujun feel free to pick this up.

**Comment by DanielQujun** [comment]
> Apologies, I completely forgot about this. @DanielQujun feel free to pick this up.

Thanks for your generous reply. After reading the design doc, this feature is prepared for Prometheus v3, so maybe it no needs to hurry, I should search more before bothering you.

BTW, I hope it is not rude to open a new PR, since I may reuse some code from you, do you mind that?

**Comment by roidelapluie** [comment]
see #12019

**Comment by roidelapluie** [comment]
This feature has been implemented in  #12019 as `scrape_config_files`. Thank you for your contribution!

**Pull request closed without merging**